pkg/vm/builtins/modulo.go

```diff
@@ -464,78 +479,122 @@ func (m *ModBuiltin) fillValue(mem *memory.Memory, inputs ModBuiltinInputs, inde
-	}
-	if err := addModBuiltinRunner.fillInputs(mem, addModBuiltinAddr, addModBuiltinInputs); err != nil {
-		return err
-	}
-	if err := addModBuiltinRunner.fillOffsets(mem, addModBuiltinInputs.offsetsPtr, nAddModsIndex, addModBuiltinInputs.n-nAddModsIndex); err != nil {
-		return err
+	if nMulMods != 0 {
+		mulModBuiltinSegment, ok := mem.FindSegmentWithBuiltin("MulMod")
+		if !ok {
+			return fmt.Errorf("MulMod builtin segment doesn't exist")
+		}
+		mulModBuiltinRunner, ok = mulModBuiltinSegment.BuiltinRunner.(*ModBuiltin)
+		if !ok {
+			return fmt.Errorf("mulModBuiltinRunner is not a ModBuiltin")
+		}
+	} else {
+		mulModBuiltinRunner = nil
 	}
 
-	mulModBuiltinInputs, err := mulModBuiltinRunner.readInputs(mem, mulModBuiltinAddr, true)
-	if err != nil {
-		return err
-	}
-	if err := mulModBuiltinRunner.fillInputs(mem, mulModBuiltinAddr, mulModBuiltinInputs); err != nil {
-		return err
-	}
-	if err := mulModBuiltinRunner.fillOffsets(mem, mulModBuiltinInputs.offsetsPtr, nMulModsIndex, mulModBuiltinInputs.n-nMulModsIndex); err != nil {
-		return err
-	}
+	var addModBuiltinInputs, mulModBuiltinInputs ModBuiltinInputs
+	var err error
 
-	addModIndex, mulModIndex := uint64(0), uint64(0)
-	for addModIndex < nAddModsIndex {
-		ok, err := addModBuiltinRunner.fillValue(mem, addModBuiltinInputs, int(addModIndex), Add)
+	if addModBuiltinRunner != nil {
+		addModBuiltinInputs, err = addModBuiltinRunner.readInputs(mem, addModInputAddress, true)
 		if err != nil {
 			return err
 		}
-		if ok {
-			addModIndex++
+		if err := addModBuiltinRunner.fillInputs(mem, addModInputAddress, addModBuiltinInputs); err != nil {
+			return err
+		}
+		if err := addModBuiltinRunner.fillOffsets(mem, addModBuiltinInputs.offsetsPtr, nAddMods, addModBuiltinInputs.n-nAddMods); err != nil {
+			return err
 		}
 	}
 
-	for mulModIndex < nMulModsIndex {
-		ok, err = mulModBuiltinRunner.fillValue(mem, mulModBuiltinInputs, int(mulModIndex), Mul)
+	if mulModBuiltinRunner != nil {
+		mulModBuiltinInputs, err = mulModBuiltinRunner.readInputs(mem, mulModInputAddress, true)
 		if err != nil {
 			return err
 		}
-		if ok {
+	}
+
+	addModIndex, mulModIndex := uint64(0), uint64(0)
+	nComputedMulGates := uint64(0)
+	for addModIndex < nAddMods || mulModIndex < nMulMods {
+		if addModIndex < nAddMods && addModBuiltinRunner != nil {
+			res, err := addModBuiltinRunner.fillValue(mem, addModBuiltinInputs, int(addModIndex), Add)
+			if err != nil {
+				return err
+			}
+			if res == 1 {
+				addModIndex++
+			}
+		}
+
+		if mulModIndex < nMulMods && mulModBuiltinRunner != nil {
+			res, err := mulModBuiltinRunner.fillValue(mem, mulModBuiltinInputs, int(mulModIndex), Mul)
+			if err != nil {
+				return err
+			}
+			if res == 0 {
+				return fmt.Errorf("MulMod builtin: Could not fill the values table")
+			}
+			if res == 2 && nComputedMulGates == 0 {
+				nComputedMulGates = mulModIndex
+			}
 			mulModIndex++
 		}
-	}
-	// POTENTIALY: add n_computed_mul_gates features in the future
```

Comment:
I see that you had to add this feature. I suppose it was related to the bug you found. Could you add a comment explaining about why is needed or what it does?

Reply:
Well no, this feature had to be implemented independently of that. Since this modifies the mulmod builtin memory segment. Though I don't understand why this was being done since there was not test for it.